File: depth_first_search/1448_count_good_nodes_in_binary_tree/1448_count_good_nodes_in_binary_tree.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func goodNodes(root *TreeNode) (goodNodeCount int) {
	getGoodNodes(root, root.Val, &goodNodeCount)
	return
}

func getGoodNodes(root *TreeNode, prev int, count *int) {
	if root == nil {
		return
	}

	if prev <= root.Val {
		*count += 1
		prev = root.Val
	}

	getGoodNodes(root.Left, prev, count)
	getGoodNodes(root.Right, prev, count)
}

File: depth_first_search/1448_count_good_nodes_in_binary_tree/1448_count_good_nodes_in_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.res = 0

        def dfs(node, curmx):
            if not node:
                return
            if node.val >= curmx:
                self.res += 1
            curmx = max(curmx, node.val)
            dfs(node.left, curmx)
            dfs(node.right, curmx)

        dfs(root, -float('inf'))
        return self.res
